doc/tkiv.3
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "App::tkiv 3"
.TH App::tkiv 3 2025-01-03 "perl v5.40.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
App::tkiv \- An image viewer in Perl::Tk based on IrfanView
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.Vb 1
\& $ iv [\-v[#]] [\-f] [\-1|\-s] [option=value ...] [dir]
\&
\& $ iv \-f imageposition=s thumsorting=size .
.Ve
.SH DESCRIPTION
.IX Header "DESCRIPTION"
\&\f(CW\*(C`iv\*(C'\fR is a pure\-perl application that was written to mimic the behavior
of the image view and manipulation program \f(CW\*(C`irfanview\*(C'\fR, which is available
as freeware on Windows.
.PP
Linux already offers many free image view and manipulation programs, but
none gave me the key\-driven approach that I liked in IrfanView, and most
left traces of what was viewed in hidden files (thumbnail previews etc).
.PP
\&\f(CW\*(C`iv\*(C'\fR is not written to be a \fIbetter\fR image viewer. It is written to
prove such an application can be written in perl and still is fast and
versatile enough to need my (own) needs.
.PP
\&\f(CW\*(C`iv\*(C'\fR is not aiming for a complete rewrite of IrfanView.
.SH "Command Line Arguments"
.IX Header "Command Line Arguments"
.IP \-v[#] 2
.IX Item "-v[#]"
Set verbose level. Default value is 0. A plain \-v without value will set
the level to 1.
.IP \-f 2
.IX Item "-f"
Start in full\-screen mode. Thumbnail view will not be full\-screen, but the
displayed images will. This overrules the default option from .ivrc files.
.IP \-s 2
.IX Item "-s"
Start the slideview on open automatically.
.IP \-1 2
.IX Item "-1"
After populating the thumbnail view, immediately display the first image
and minimize the thumbnail view.
.SS Options
.IX Subsection "Options"
Options are case insensitive. They are read from /etc/iv.rc, /etc/.ivrc,
~/iv.rc, ~/.ivrc, ./iv.rc, ./.ivrc, image_dir/iv.rc and image_dir/.ivrc
(in that order if available).
.PP
Not (yet) local per dir, so the last one read stays active and overrules
the previous settings.
.PP
.Vb 2
\&    thumbSize           => 80,          # in pixels
\&        size of the thumbnails in the thumbview. default 80x80
\&
\&    thumbRows           => 5,
\&        number of rows in the thumbview. default 5
\&
\&    thumbPosition       => "se",
\&        position of the thumbview on the screen. default se (bottom right)
\&        possible values: n, ne, e, se, s, sw, w, nw, c
\&
\&    thumbRefresh        => 1,
\&        when deleting images, the thumb preview is also deleted, and all
\&        thumbs then have to be re\-displayed in order to fill the gap. When
\&        deleting images out of a long sequence, this will take a long time.
\&        Disabling the refresh will make the cleanup procedure go faster,
\&        but will cause errors/warnings when you re\-visit the thumb that
\&        has been deleted
\&
\&    thumbSorting        => "default",
\&        defines the ordering of the thumbnails in the thumbview.
\&
\&        default : on the numeric part of the image name, then on the full name
\&        caseless: on the lower\-cased image name
\&        date    : on mtime
\&        size    : on (file) size
\&        random  : random
\&
\&    thumbSortOrder      => "ascending",
\&        ordering of the thumbs. default ascending. possible values ascending
\&        or descending
\&
\&    imagePosition       => "nw",
\&        position of the image window. default nw (top\-left)
\&        possible values: n, ne, e, se, s, sw, w, nw, c
\&
\&    imageDir            => ".",
\&        default image dir (current folder)
\&
\&    slideShowDelay      => 1500,        # in milliseconds
\&        in slide shows, define the display time for each image. default: 1500 ms
\&
\&    slidePosition       => "c",
\&        position of the full image in slide\-show mode. default c (center)
\&        possible values: n, ne, e, se, s, sw, w, nw, c
\&
\&    slideFull           => 0,
\&        display images auto\-fit in slide\-show mode. default 0 (off)
\&
\&    slideCover          => 0,
\&        EXPERIMENTAL
\&        cover the background when in slideshow mode
\&
\&    titleDirs           => 0
\&        The number of dir levels (counted from the end) that should be included
\&        in the image title
\&
\&    titleIndex          => 0
\&        display the index of the image in the image title. (default off)
\&        When on, the title of the image window is extended with " n/N", where
\&        n is the index and N is the total number of images in the current
\&        thumb\-view
\&
\&    maxX                => 9999,
\&        maximum horizontal resolution (some size quests are unreliable)
\&
\&    maxY                => 9999,
\&        maximum vertical resolution (some size quests are unreliable)
\&
\&    smallFont           => "{Liberation Mono} 8",",
\&        define the font for labels
\&
\&    selectionFont       => "{Liberation Sans} 5",
\&        the font used to display information with the selection rectangle
\&
\&    selectionColor      => "Yellow",
\&        the color used when drawing selection rectangles
\&
\&    confirmDelete       => 1,
\&        ask for confirmation when deleting images. default 1 (true)
\&
\&    removeTarget        => 0,
\&        remove target folder when it is empty. default 0 (false)
\&
\&    imageFull           => 0,
\&        show images full\-screen. default 0 (off). Full\-screen resizes the
\&        image to automatically fit the screen height or screen\-width when
\&        maintaining aspect ratio. Also sometimes called auto\-fit.
\&
\&    lastFirstNext       => 0,
\&        If set will cause keys_nextpic on the last pic to open the first
\&        pic of the next set as if keys_firstnext was used
\&
\&    showExifInfo        => 0,
\&        show a summary of the exif info of the image on the image
\&
\&    exifInfoColor       => "Blue",
\&        the color if the exif info text on the image for showexifinfo
\&
\&    decoration          => 1,
\&        show the screen control elements. Default 1 (on). Some window
\&        managers support full control, either by keys or mouse, even when
\&        the borders and title bar are gone. Some refuse to do scroll bars
\&        and such when the decoration is disabled. YMMV
\&
\&    scrollSpeed         => 3,
\&        The default scroll increments for Tk::Canvas is a bit too course
\&        for fine\-grained scrolling. A low value enables precise positioning
\&        with the scroll wheel, high values scroll fast.
\&
\&    dirTreeStartPos     => 0.,
\&        The initial position of the dir tree view relative from the top.
\&        The default is to show the top of the tree if there is more than
\&        the current window to be viewed and a scrollbar is shown. Setting
\&        dirTreeStartPos to 1. will show the bottom of the tree on startup.
\&        A value of .5 will show the middle.
.Ve
.SS "Key bindings"
.IX Subsection "Key bindings"
Key bindings. Most are the same as the windows program IrfanView, after
which iv was initially modeled
.PP
.Vb 2
\&                        Thumb view              Image view
\&  keys_quit
\&
\&    q ESC               Quit program            Close image,
\&                                                return to thumbs
\&
\&  keys_quit_all
\&
\&    Q Control\-q         Quit program            Quit program
\&
\&  keys_options
\&
\&    o                   Open option window      \-
\&
\&  keys_firstpic
\&
\&    0 1 a               Open first pic          id
\&
\&  keys_prevpic
\&
\&    Left Up BS          Prev pic (round robin)  id
\&
\&  keys_nextpic
\&
\&    Space Right Down    Next pic                id
\&                         When LastFirstNext is false (default), this
\&                         will round\-robin to the first pic, otherwise
\&                         it will open the first pic in the next set
\&
\&  keys_lastpic
\&
\&    9 z                 Open last pic           id
\&
\&  keys_firstnext
\&
\&    v                   \-                       Open first pic in next folder
\&
\&  keys_firstprev
\&
\&    ^                   \-                       Open first pic in prev folder
\&
\&  keys_fullscreen
\&
\&    f F11               \-                       Toggle auto\-fit
\&
\&  keys_fitwidth
\&
\&    b                   \-                       Zoom to fit screen\-width
\&
\&  keys_fitheight
\&
\&    h                   \-                       Zoom to fit screen\-height
\&
\&  keys_origsize
\&
\&    o                   \-                       Display in original size
\&
\&  keys_full_toggle
\&
\&    Alt\-f               \-                       Toggle auto\-fit and remember
\&                                                for next sets
\&
\&  keys_full_rc
\&
\&    F                   \-                       Switch to auto\-fit and save
\&                                                in ./.ivrc
\&
\&  keys_rotleft
\&
\&    l                   \-                       Rotate image left (90°
\&                                                anti\-clockwise)
\&
\&  keys_rotexifl
\&
\&    L                   \-                       Rotate image left (90° anti\-
\&                                                clockwise) in the EXIF info,
\&                                                and store the change in the
\&                                                image file
\&
\&  keys_rotright
\&
\&    r                   \-                       Rotate image right (90°
\&                                                clockwise)
\&
\&  keys_rotexifr
\&
\&    R                   \-                       Rotate image right (90°
\&                                                clockwise) in the EXIF info,
\&                                                and store the change in the
\&                                                image file
\&  keys_zoomin
\&
\&    plus                \-                       Zoom in with 20% steps.
\&
\&  keys_zoomout
\&
\&    minus               \-                       Zoom out with 20% steps.
\&
\&  keys_delete
\&
\&    Delete              \-                       Delete image from disk
\&
\&  keys_slideshow
\&
\&    w s                 Start slideshow         id
\&
\&  keys_exif
\&
\&    i                   \-                       Show EXIF info
\&
\&  keys_exifinfo
\&
\&    I                   \-                       Show EXIF info summary on image
\&
\&  keys_decoration
\&
\&    d                   \-                       Remove decoration. Unreliable,
\&                                                you might loose control or key
\&                                                bindings.
\&
\&  keys_focusthumbs
\&
\&    t                   \-                       Restore and focus thumbnails
\&
\&  keys_imgpos_nw
\&
\&    Alt\-u               \-                       Set image position to nw
\&
\&  keys_imgpos_n
\&
\&    Alt\-i               \-                       Set image position to n
\&
\&  keys_imgpos_ne
\&
\&    Alt\-o               \-                       Set image position to ne
\&
\&  keys_imgpos_e
\&
\&    Alt\-l               \-                       Set image position to e
\&
\&  keys_imgpos_se
\&
\&    Alt\-period          \-                       Set image position to se
\&
\&  keys_imgpos_s
\&
\&    Alt\-comma           \-                       Set image position to s
\&
\&  keys_imgpos_sw
\&
\&    Alt\-m               \-                       Set image position to sw
\&
\&  keys_imgpos_w
\&
\&    Alt\-j               \-                       Set image position to w
\&
\&  keys_imgpos_c
\&
\&    Alt\-k               \-                       Set image position to c
\&
\&
\&  keys_crop
\&
\&    Control\-y           \-                       Crop image to selection box
.Ve
.PP
Zoom factors are limited to 1 2 3 4 5 7 9 11 14 17 21 26 32 39 47 57 69 83 100
120 144 172 206 247 296 355 426 511 613 735 882 1058 1269 1522 1826 2191 2629
3154 3784 4540 5448 6537 7844 9412 and 11300
.SH "EXIF INFO"
.IX Header "EXIF INFO"
\&...
.SH EXAMPLES
.IX Header "EXAMPLES"
\&...
.SH DIAGNOSTICS
.IX Header "DIAGNOSTICS"
\&...
.SH "BUGS and CAVEATS"
.IX Header "BUGS and CAVEATS"
For manipulation (resizing, rotation, ...) the external command \f(CW\*(C`convert\*(C'\fR
from ImageMagick is used.
.SH TODO
.IX Header "TODO"
.IP "save/load from .ivrc buttons on option window" 2
.IX Item "save/load from .ivrc buttons on option window"
.PD 0
.IP Slideshow 2
.IX Item "Slideshow"
.PD
.Vb 5
\& \- behavior: location, dir depth, cycling
\& \- randomness, slide lists, full screen background (no decoration)
\& \- playlist
\& \- loop control
\& \- Auto\-sense image load time for slideshows
.Ve
.IP "Image manipulation" 2
.IX Item "Image manipulation"
.Vb 3
\& \- selection less picky
\& \- selection from zoom other than original
\& \- Save, Save as ...
.Ve
.IP "Titles and decoration behavior" 2
.IX Item "Titles and decoration behavior"
.Vb 2
\& \- adjust height/width of screen\-fit images to decoration
\&   (I just cannot get $iv\->overrideredirect (1) to work as I want)
.Ve
.IP "Tree view" 2
.IX Item "Tree view"
.Vb 4
\& \- Hide dirs above dt root
\& \- Allow a set of dirs from the command line
\& \- Make pirstnext and firstprev look in the original folder if the
\&   image folder actually been viewed is a symlink in the parent folder
.Ve
.IP Animation 2
.IX Item "Animation"
.Vb 1
\& \- use Tk::Animation for animated gif\*(Aqs
.Ve
.IP Menu\*(Aqs 2
.IX Item "Menu's"
.PD 0
.IP Documentation 2
.IX Item "Documentation"
.IP Portability 2
.IX Item "Portability"
.PD
.Vb 1
\& \- make image format\*(Aqs optional (TIFF, NEF, ...)
.Ve
.SH "SEE ALSO"
.IX Header "SEE ALSO"
perl, Tk, Tk::JPEG, Tk::PNG, Tk::TIFF, Tk::Bitmap,
Tk::Pixmap, Tk::Animation, Image::ExifTool, Image::Size,
Image::Info and ImageMagick.
.SH WARRANTY
.IX Header "WARRANTY"
The fact that I use it on my own picture sets is by no means a guarantee
that the software is without bugs. Use with care, and make backups of all
pictures you care about before experimenting.
.SH AUTHOR
.IX Header "AUTHOR"
H.Merijn Brand \fI<h.m.brand@xs4all.nl>\fR wrote this for his own
personal use, but was asked to make it publicly available as application.
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
Copyright (C) 2004\-2025 H.Merijn Brand
.PP
This software is free; you can redistribute it and/or modify
it under the same terms as Perl itself.


ChangeLog
0.134 - 03 Jan 2025, H.Merijn Brand
    * More specific version declarations based on known CVE's
    * Pre-grenerated documentation
    * It's 2024
    * It's 2025

0.133 - 06 Jan 2023, H.Merijn Brand
    * It's 2023

0.132 - 01 Jan 2022, H.Merijn Brand
    * It's 2021 - no changes in 2021
    * It's 2022

0.131 - 17 Dec 2020, H.Merijn Brand
    * Checked with perl-5.24.0 .. 5.30.0
    * It's 2017
    * It's 2018
    * It's 2019
    * Add cpanfile
    * It's 2020

0.130 - 02 Jan 2016, H.Merijn Brand
    * Add a third line to in-image exif info showing device/camera
    * Add CONTRIBUTING.md
    * Use File::Temp for temporary files (esp. for Windows users)
    * It's 2016

0.129 - 11 Feb 2015, H.Merijn Brand
    * Remove defined-or test from Makefile.PL
    * Remove perl recommendation from META as it breaks cpan clients
    * Move repo to github

0.128 - 02 Jan 2015, H.Merijn Brand
    * Require Image::ExifInfo (was optional) and META info for rest

0.127 - 01 Jan 2015, H.Merijn Brand
    * Update copyright to 2014
    * Open images under eval to skip invalid files
    * Update copyright to 2015

0.126 - 12 May 2013, H.Merijn Brand
    * Update copyright to 2013
    * Declare UTF-8 in documentation
    * Support buttons <5> and <6> for horizontal scrolling

0.125 - 11 Jan 2012, H.Merijn Brand
    * Update copyright to 2012

0.124 - 07 Sep 2011, H.Merijn Brand
    * NAME / DISTNAME in Makefile.PL

0.123 - 07 Sep 2011, H.Merijn Brand
    * Safeguard on failing exiftran
    * More cross-checks for META data
    * Follow original path when targets are sym-linked instead of realpath
    * Don't expand thumbs when resizing main window
    * Make firstprev work
    * Introduce dirTreeStartPos

0.122 - 11 Aug 2010, H.Merijn Brand
    * Allow iv.rc as well
    * Introduce keys_firstprev
    * Spell checking

0.121 - 03 Jun 2010, H.Merijn Brand
    * Introduce keys_focusthumbs
    * Introduce titleIndex
    * Update examples/.ivrc

0.120 - 12 Apr 2010, H.Merijn Brand
    * Introduce keys_full_toggle
    * Introduce lastFirstNext option
    * Hugely speed up user experience on firstNext and lastFirstNext
    * Dropped YAML spec to 1.0
    * Moved pod to App::tkiv

0.116 - 15 Dec 2009, H.Merijn Brand
    * Updated META.yml to meet most recent specs (optional_features = map)

0.115 - 15 Dec 2009, H.Merijn Brand
    * Correct name in META.yml

0.114 - 31 Oct 2009, H.Merijn Brand
    * META.yml declaration fix

0.113 - 31 Oct 2009, H.Merijn Brand
    * Correct wrong eval { use ... }} checking of optional modules
    * system call now takes list to prevent quote detection
    * Delayed destroy of previous pic display to ease transitions (thanks JV)
    * Introduce ScrollSpeed option
    * Introduce ThumbRefresh option
    * Use both possible methods to detect screen-size and use largest values
    * Introduce -1 command line option
    * Update and expand documentation
    * Introduce keys_quit_all
    * Add missing start with last pic key binding
    * Introduce keys_firstnext
    * Introduce titledirs
    * Bad images are now skipped instead of causing a die

0.112 - 11 Aug 2009, H.Merijn Brand
    * Make X11::Protocol optional, and use $mw->screenwidth

0.111 - 07 Aug 2009, H.Merijn Brand
    * Canvas has native restrictive scroll (Wheel, Alt-Wheel and Control-Wheel),
      no more Scrolled subwidget (still need working key-bindings)

0.110 - 28 Jul 2009, H.Merijn Brand
    * Add showexifinfo

0.103 - 19 Jul 2009, H.Merijn Brand
    * Make slidecover work (a bit)

0.102 - 05 Jul 2009, H.Merijn Brand
    * removed unused slideshowloop attrib

0.101 - 27 Apr 2009, H.Merijn Brand
    * selectionFont & selectionColor attributes
    * better deal with maxx (Xinerama dual screens)

0.100 - 05 Mar 2009, H.Merijn Brand
    * Cropping/selection (switch from Label to Canvas to display full images)
    * Data::Peek instead of Data::Dumper


0.051 - 30 Jan 2009, H.Merijn Brand
    * Die if PhotoXY fails

0.050 - 09 Jul 2008, H.Merijn Brand
    * Switch font syntax

0.049 - 08 Jul 2008, H.Merijn Brand
    * Exif Info window now expands (top level pack i.o. grid)

0.048 - 25 Nov 2007, H.Merijn Brand
    * Added NEF (a TIFF variant)

0.047 - 05 Nov 2007, H.Merijn Brand
    * Small changes in preparation of App::tkiv (defined-or)

0.046 - 29 Oct 2007, H.Merijn Brand
    * Added Tk::TIFF which works fine with 804.027_501

0.045 - 05 Oct 2007, H.Merijn Brand
    * ...

0.044 - 25 Sep 2007, H.Merijn Brand
    * Upgrade window titles to UTF-8
    * Some perlcritic cleanups (perlcritic also evolves)

0.043 - 17 Sep 2007, H.Merijn Brand
    * Reset update status if folder deleted
    * Remember the refreshed folder in refreshing

0.042 - 10 Sep 2007, H.Merijn Brand
    * Prevent update while update

0.041 - 09 Sep 2007, H.Merijn Brand
    * Some perl::critic cleanups

0.040 - 03 Sep 2007, H.Merijn Brand
    * Cleaned up slideshow somewhat. added -s

0.039 - 30 Aug 2007, H.Merijn Brand
    * Added random sorting for images

0.038 - 18 Aug 2007, H.Merijn Brand
    * Reverse is alias for descending for ThumbSortOrder

0.037 - 26 Jun 2007, H.Merijn Brand
    * Support for Image::Size; better parsing for identify

0.036 - 22 Jun 2007, H.Merijn Brand
    * Added key bindings to set image position

0.035 - 11 Feb 2007, H.Merijn Brand
    * Added fit-to-height (h), fit-to-width (b) and orig-size (o)
    * Added changelog, delete folder if empty, fix key-F

0.034 - 02 Jan 2007, H.Merijn Brand
    * Rotation saves Orientation *and* Rotation

0.033 - 16 Feb 2006, H.Merijn Brand
    * Skip un-identify-able pics, up- and down keys in treeview

0.032 - 06 Feb 2006, H.Merijn Brand
    * Show legal values for alt options in usage

0.031 - 10 Dec 2005, H.Merijn Brand
    * show options, default-keys, more TODO, $opt_v, Image::Info,
      background/backdrop, slide-show options, first stab at
      animated pics, key F stores fullscreen on current dir,

0.025 - 15 Aug 2005, H.Merijn Brand

lib/App/tkiv.pm
#!/usr/bin/perl

package App::tkiv;

use strict;
use warnings;

our $VERSION = "0.134";

1;

__END__

=encoding UTF-8

=head1 NAME

App::tkiv - An image viewer in Perl::Tk based on IrfanView

=head1 SYNOPSIS

 $ iv [-v[#]] [-f] [-1|-s] [option=value ...] [dir]

 $ iv -f imageposition=s thumsorting=size .

=head1 DESCRIPTION

C<iv> is a pure-perl application that was written to mimic the behavior
of the image view and manipulation program C<irfanview>, which is available
as freeware on Windows.

Linux already offers many free image view and manipulation programs, but
none gave me the key-driven approach that I liked in IrfanView, and most
left traces of what was viewed in hidden files (thumbnail previews etc).

C<iv> is not written to be a I<better> image viewer. It is written to
prove such an application can be written in perl and still is fast and
versatile enough to need my (own) needs.

C<iv> is not aiming for a complete rewrite of IrfanView.

=head1 Command Line Arguments

=over 2

=item -v[#]

Set verbose level. Default value is 0. A plain -v without value will set
the level to 1.

=item -f

Start in full-screen mode. Thumbnail view will not be full-screen, but the
displayed images will. This overrules the default option from .ivrc files.

=item -s

Start the slideview on open automatically.

=item -1

After populating the thumbnail view, immediately display the first image
and minimize the thumbnail view.

=back

=head2 Options

Options are case insensitive. They are read from /etc/iv.rc, /etc/.ivrc,
~/iv.rc, ~/.ivrc, ./iv.rc, ./.ivrc, image_dir/iv.rc and image_dir/.ivrc
(in that order if available).

Not (yet) local per dir, so the last one read stays active and overrules
the previous settings.

    thumbSize		=> 80,		# in pixels
	size of the thumbnails in the thumbview. default 80x80

    thumbRows		=> 5,
	number of rows in the thumbview. default 5

    thumbPosition	=> "se",
	position of the thumbview on the screen. default se (bottom right)
	possible values: n, ne, e, se, s, sw, w, nw, c

    thumbRefresh	=> 1,
	when deleting images, the thumb preview is also deleted, and all
	thumbs then have to be re-displayed in order to fill the gap. When
	deleting images out of a long sequence, this will take a long time.
	Disabling the refresh will make the cleanup procedure go faster,
	but will cause errors/warnings when you re-visit the thumb that
	has been deleted

    thumbSorting	=> "default",
	defines the ordering of the thumbnails in the thumbview.

	default	: on the numeric part of the image name, then on the full name
	caseless: on the lower-cased image name
	date	: on mtime
	size	: on (file) size
	random	: random

    thumbSortOrder	=> "ascending",
	ordering of the thumbs. default ascending. possible values ascending
	or descending

    imagePosition	=> "nw",
	position of the image window. default nw (top-left)
	possible values: n, ne, e, se, s, sw, w, nw, c

    imageDir		=> ".",
	default image dir (current folder)

    slideShowDelay	=> 1500,	# in milliseconds
	in slide shows, define the display time for each image. default: 1500 ms

    slidePosition	=> "c",
	position of the full image in slide-show mode. default c (center)
	possible values: n, ne, e, se, s, sw, w, nw, c

    slideFull		=> 0,
	display images auto-fit in slide-show mode. default 0 (off)

    slideCover		=> 0,
	EXPERIMENTAL
	cover the background when in slideshow mode

    titleDirs		=> 0
	The number of dir levels (counted from the end) that should be included
	in the image title

    titleIndex		=> 0
	display the index of the image in the image title. (default off)
	When on, the title of the image window is extended with " n/N", where
	n is the index and N is the total number of images in the current
	thumb-view

    maxX		=> 9999,
	maximum horizontal resolution (some size quests are unreliable)

    maxY		=> 9999,
	maximum vertical resolution (some size quests are unreliable)

    smallFont		=> "{Liberation Mono} 8",",
	define the font for labels

    selectionFont	=> "{Liberation Sans} 5",
	the font used to display information with the selection rectangle

    selectionColor	=> "Yellow",
	the color used when drawing selection rectangles

    confirmDelete	=> 1,
	ask for confirmation when deleting images. default 1 (true)

    removeTarget	=> 0,
	remove target folder when it is empty. default 0 (false)

    imageFull		=> 0,
	show images full-screen. default 0 (off). Full-screen resizes the
	image to automatically fit the screen height or screen-width when
	maintaining aspect ratio. Also sometimes called auto-fit.

    lastFirstNext	=> 0,
	If set will cause keys_nextpic on the last pic to open the first
	pic of the next set as if keys_firstnext was used

    showExifInfo	=> 0,
	show a summary of the exif info of the image on the image

    exifInfoColor	=> "Blue",
	the color if the exif info text on the image for showexifinfo

    decoration		=> 1,
	show the screen control elements. Default 1 (on). Some window
	managers support full control, either by keys or mouse, even when
	the borders and title bar are gone. Some refuse to do scroll bars
	and such when the decoration is disabled. YMMV

    scrollSpeed		=> 3,
	The default scroll increments for Tk::Canvas is a bit too course
	for fine-grained scrolling. A low value enables precise positioning
	with the scroll wheel, high values scroll fast.

    dirTreeStartPos	=> 0.,
	The initial position of the dir tree view relative from the top.
	The default is to show the top of the tree if there is more than
	the current window to be viewed and a scrollbar is shown. Setting
	dirTreeStartPos to 1. will show the bottom of the tree on startup.
	A value of .5 will show the middle.

=head2 Key bindings

Key bindings. Most are the same as the windows program IrfanView, after
which iv was initially modeled

			Thumb view		Image view
  keys_quit

    q ESC		Quit program		Close image,
						return to thumbs

  keys_quit_all

    Q Control-q		Quit program		Quit program

  keys_options

    o			Open option window	-

  keys_firstpic

    0 1 a		Open first pic		id

  keys_prevpic

    Left Up BS		Prev pic (round robin)	id

  keys_nextpic

    Space Right Down	Next pic		id
			 When LastFirstNext is false (default), this
			 will round-robin to the first pic, otherwise
			 it will open the first pic in the next set

  keys_lastpic

    9 z			Open last pic		id

  keys_firstnext

    v			-			Open first pic in next folder

  keys_firstprev

    ^			-			Open first pic in prev folder

  keys_fullscreen

    f F11		-			Toggle auto-fit

  keys_fitwidth

    b			-			Zoom to fit screen-width

  keys_fitheight

    h			-			Zoom to fit screen-height

  keys_origsize

    o			-			Display in original size

  keys_full_toggle

    Alt-f		-			Toggle auto-fit and remember
						for next sets

  keys_full_rc

    F			-			Switch to auto-fit and save
						in ./.ivrc

  keys_rotleft

    l			-			Rotate image left (90°
						anti-clockwise)

  keys_rotexifl

    L			-			Rotate image left (90° anti-
						clockwise) in the EXIF info,
						and store the change in the
						image file

  keys_rotright

    r			-			Rotate image right (90°
						clockwise)

  keys_rotexifr

    R			-			Rotate image right (90°
						clockwise) in the EXIF info,
						and store the change in the
						image file
  keys_zoomin

    plus		-			Zoom in with 20% steps.

  keys_zoomout

    minus		-			Zoom out with 20% steps.

  keys_delete

    Delete		-			Delete image from disk

  keys_slideshow

    w s			Start slideshow		id

  keys_exif

    i			-			Show EXIF info

  keys_exifinfo

    I			-			Show EXIF info summary on image

  keys_decoration

    d			-			Remove decoration. Unreliable,
						you might loose control or key
						bindings.

  keys_focusthumbs

    t			-			Restore and focus thumbnails

  keys_imgpos_nw

    Alt-u		-			Set image position to nw

  keys_imgpos_n

    Alt-i		-			Set image position to n

  keys_imgpos_ne

    Alt-o		-			Set image position to ne

  keys_imgpos_e

    Alt-l		-			Set image position to e

  keys_imgpos_se

    Alt-period		-			Set image position to se

  keys_imgpos_s

    Alt-comma		-			Set image position to s

  keys_imgpos_sw

    Alt-m		-			Set image position to sw

  keys_imgpos_w

    Alt-j		-			Set image position to w

  keys_imgpos_c

    Alt-k		-			Set image position to c


  keys_crop

    Control-y		-			Crop image to selection box

Zoom factors are limited to 1 2 3 4 5 7 9 11 14 17 21 26 32 39 47 57 69 83 100
120 144 172 206 247 296 355 426 511 613 735 882 1058 1269 1522 1826 2191 2629
3154 3784 4540 5448 6537 7844 9412 and 11300

=head1 EXIF INFO

...

=head1 EXAMPLES

...

=head1 DIAGNOSTICS

...

=head1 BUGS and CAVEATS

For manipulation (resizing, rotation, ...) the external command C<convert>
from ImageMagick is used.

=head1 TODO

=over 2

=item save/load from .ivrc buttons on option window

=item Slideshow

 - behavior: location, dir depth, cycling
 - randomness, slide lists, full screen background (no decoration)
 - playlist
 - loop control
 - Auto-sense image load time for slideshows

=item Image manipulation

 - selection less picky
 - selection from zoom other than original
 - Save, Save as ...

=item Titles and decoration behavior

 - adjust height/width of screen-fit images to decoration
   (I just cannot get $iv->overrideredirect (1) to work as I want)

=item Tree view

 - Hide dirs above dt root
 - Allow a set of dirs from the command line
 - Make pirstnext and firstprev look in the original folder if the
   image folder actually been viewed is a symlink in the parent folder

=item Animation

 - use Tk::Animation for animated gif's

=item Menu's

=item Documentation

=item Portability

 - make image format's optional (TIFF, NEF, ...)

=back

=head1 SEE ALSO

L<perl>, L<Tk>, L<Tk::JPEG>, L<Tk::PNG>, L<Tk::TIFF>, L<Tk::Bitmap>,
L<Tk::Pixmap>, L<Tk::Animation>, L<Image::ExifTool>, L<Image::Size>,
L<Image::Info> and ImageMagick.

=head1 WARRANTY

The fact that I use it on my own picture sets is by no means a guarantee
that the software is without bugs. Use with care, and make backups of all
pictures you care about before experimenting.

=head1 AUTHOR

H.Merijn Brand F<E<lt>h.m.brand@xs4all.nlE<gt>> wrote this for his own
personal use, but was asked to make it publicly available as application.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004-2024 H.Merijn Brand

This software is free; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
